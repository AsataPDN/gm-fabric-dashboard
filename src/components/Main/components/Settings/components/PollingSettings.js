import _ from "lodash";
import { PropTypes } from "prop-types";
import React, { Component } from "react";
import InputRange from "react-input-range";

import Button from "components/Button";
import LayoutSection from "components/LayoutSection";

import PollIcon from "images/icons/poll.svg";

/**
 * Control to start/stop polling and change the polling rate
 * Styled to resemble a Readout and intended to be a child of SettingsGrid
 */
class PollingSettings extends Component {
  static propTypes = {
    changePollingInterval: PropTypes.func.isRequired,
    interval: PropTypes.number.isRequired,
    isPolling: PropTypes.bool.isRequired,
    startPolling: PropTypes.func.isRequired,
    stopPolling: PropTypes.func.isRequired,
    title: PropTypes.string.isRequired
  };

  // Use local state to have a "loosely" controlled component whereby the slider
  // slides smoothly and changes to Redux are debounced.
  state = {
    localInterval: this.props.interval / 1000,
    debouncedSetInterval: _.debounce(this.props.changePollingInterval, 1000)
  };

  render() {
    const { isPolling, stopPolling, startPolling, title } = this.props;
    const buttonGlyph = isPolling ? "Pause" : "Play";
    const buttonLabel = isPolling ? "Pause Updates" : "Resume Updates";
    return (
      <LayoutSection title={title} icon={PollIcon} flex>
        <div
          className="control-group control-group-polling-start-stop"
          id={`ctrl-btn-${title}`}
        >
          <Button
            clickAction={() => {
              if (isPolling) {
                stopPolling();
              } else {
                startPolling();
              }
            }}
            glyph={buttonGlyph}
            label={buttonLabel}
            orientation={"vertical"}
            outline={"raised"}
            tabIndex={0}
            type={"polling"}
            style={{
              width: "100%",
              height: "100%",
              position: "relative"
            }}
          />
        </div>

        <div
          className="control-group control-group-polling-interval"
          id={`ctrl-slider-${title}`}
        >
          <InputRange
            aria-labelledby="polling interval-name"
            maxValue={120}
            minValue={5}
            onChange={value => {
              this.setState({ localInterval: value });
              this.state.debouncedSetInterval(value * 1000);
            }}
            value={this.state.localInterval}
          />
          <span className="label" id={`interval-name-${title}`}>
            {"Updates Per Minute"}
          </span>
        </div>
      </LayoutSection>
    );
  }
}

export default PollingSettings;
