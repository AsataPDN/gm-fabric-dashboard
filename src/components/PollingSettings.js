import React, { Component, PropTypes } from 'react';
import InputRange from 'react-input-range';
import _ from 'lodash';
import { Actions } from 'jumpstate';

class PollingSettings extends Component {
  // This state is solely to allow smooth animation on the slider and the ability to debounce the 
  // dispatch of the setInterval action creator.

  static propTypes = {
    interval: PropTypes.number.isRequired,
    isPolling: PropTypes.bool.isRequired
  }

  state = {
    localInterval: this.props.interval / 1000,
    debouncedSetInterval: _.debounce(Actions.setInterval, 1000)
  }
  render() {
    const buttonClass = this.props.isPolling ? 'uk-button uk-button-danger' : 'uk-button uk-button-primary';
    return (
      <div className="uk-card uk-card-default uk-card-body">
        <h3 className="uk-card-title">Nonfunctional Dummy Settings</h3>
        <button
          className={buttonClass}
          onClick={() => Actions.togglePolling()}
        >{this.props.isPolling ? 'Stop Polling' : 'Start Polling'}
        </button>
        <h4>Polling Interval (seconds)</h4>
        <InputRange
          maxValue={120}
          minValue={5}
          onChange={value => {
            this.setState({ localInterval: value });
            this.state.debouncedSetInterval(value * 1000);
          }}
          value={this.state.localInterval}
        />
      </div>
    );
  }
}

export default PollingSettings;
