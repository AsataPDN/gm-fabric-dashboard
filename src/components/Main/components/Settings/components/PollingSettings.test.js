import React from "react";
import { mount } from "enzyme";
import renderer from "react-test-renderer";
import PollingSettings from "./PollingSettings";

let wrapper;
const mockChangePollingInterval = jest.fn();
const mockStartPolling = jest.fn();
const mockStopPolling = jest.fn();
mockStartPolling.mockReturnValue("Start Polling Function");
mockStopPolling.mockReturnValue("Stop Polling Function");

let mockProps = {
  changePollingInterval: mockChangePollingInterval,
  interval: 5000,
  isPolling: true,
  startPolling: mockStartPolling,
  stopPolling: mockStopPolling,
  title: "Instance Polling",
  glyph: "Poll"
};

describe("PollingSettings component", () => {
  beforeEach(function() {
    mockStartPolling.mockClear();
    mockStopPolling.mockClear();
    wrapper = mount(<PollingSettings {...mockProps} />);
  });

  test("Matches the snapshot", () => {
    const PollingSettingsComponent = renderer
      .create(<PollingSettings {...mockProps} />)
      .toJSON();
    expect(PollingSettingsComponent).toMatchSnapshot();
  });

  test("Resume / Pause button component is rendered", () => {
    expect(wrapper.find("button").length).toBe(1);
  });

  test("If isPolling prop is set to 'true', button displays 'Pause Updates'", () => {
    wrapper.setProps({ isPolling: true });
    expect(wrapper.find("button").props().title).toBe("Pause Updates");
  });

  test("If isPolling prop is set to 'true', and button is clicked, the function that is called is Stop Polling", () => {
    wrapper.setProps({ isPolling: true });
    wrapper.find("button").simulate("click", { preventDefault() {} });
    expect(mockStopPolling).toBeCalled();
  });

  test("If isPolling prop is set to 'false', the PollingSettings Button displays 'Resume Updates'", () => {
    wrapper.setProps({ isPolling: false });
    expect(wrapper.find("button").props().title).toBe("Resume Updates");
  });

  test("If isPolling prop is set to 'false', and button is clicked, the function that is called is Start Polling", () => {
    wrapper.setProps({ isPolling: false });
    wrapper.find("button").simulate("click", { preventDefault() {} });
    expect(mockStartPolling).toBeCalled();
  });

  test("Glyph is rendered as part of the resume/ pause button", () => {
    expect(
      wrapper
        .find("button")
        .find("g")
        .at(1)
        .hasClass("glyph")
    ).toBe(true);
  });

  test("LayoutSection component is rendered by making sure header tag contains 'Instance Polling' text", () => {
    expect(wrapper.find("LayoutSection")).toHaveLength(1);
  });

  test("InputRange component is getting rendered", () => {
    expect(wrapper.find(".input-range").length).toBe(1);
  });

  test("InputRange value depends on the initial state of the Polling Settings component", () => {
    mockProps.interval = 99000;
    const wrapper = mount(<PollingSettings {...mockProps} />);
    expect(wrapper.find("InputRange").props().value).toBe(99);
  });
});
