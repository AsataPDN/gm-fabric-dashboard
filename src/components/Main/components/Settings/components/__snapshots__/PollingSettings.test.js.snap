// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PollingSettings component Matches the snapshot 1`] = `
<section
  className="sc-htpNat gOoNtI"
>
  <header
    className="sc-bxivhb cnFCnx"
  >
    <span
      className="sc-EHOje bfNwRO"
    >
      <svg
        aria-labelledby="ariaLabelledby"
        className="sc-bwzfXH flKpft"
        version="1.1"
        viewBox="0 0 24 24"
        xmlns="http://www.w3.org/2000/svg"
        xmlnsXlink="http://www.w3.org/1999/xlink"
      >
        <g
          className="sc-bdVaJa iPrjqu"
          fill="currentColor"
          title=""
        >
          <title>
            
          </title>
          <g
            className="glyph"
            fill="currentColor"
            transform="translate(0 0) scale(1)"
          >
            <path
              d="M14.9840176,4.61673196 C17.8864554,5.80181797 19.9317664,8.65279612 19.9317664,11.9814822 C19.9317664,16.3731065 16.3716515,19.9332214 11.9800272,19.9332214 C7.58840298,19.9332214 4.02828809,16.3731065 4.02828809,11.9814822 C4.02828809,8.78101943 5.91906039,6.02217015 8.64450856,4.76103084 L5.27758564,4.46646325 L5.39309226,3.14621652 L10.8941203,3.62749411 L10.4128427,9.12852216 L9.09259598,9.01301553 L9.32170102,6.39433295 C7.22022775,7.39505768 5.79534124,9.54048188 5.79534124,11.9814822 C5.79534124,15.39719 8.56431948,18.1661683 11.9800272,18.1661683 C15.395735,18.1661683 18.1647133,15.39719 18.1647133,11.9814822 C18.1647133,9.69401394 16.9113633,7.64421418 14.9840176,6.57297136 L14.9840176,8.00561267 L13.7470804,8.00561267 L13.7470804,3.14621652 L14.9840176,3.14621652 L14.9840176,4.61673196 L14.9840176,4.61673196 Z"
              id="Combined-Shape"
            />
          </g>
        </g>
      </svg>
    </span>
    <h3
      className="sc-bZQynM hEJNNX"
    >
      Instance Polling
    </h3>
  </header>
  <div
    className="sc-ifAKCX iLlrDi"
  >
    <div
      className="sc-gzVnrw hLRaYR"
    >
      <button
        className="btn btn-type-polling btn-outline-raised btn-orientation-vertical"
        disabled={undefined}
        onClick={[Function]}
        style={
          Object {
            "height": "100%",
            "position": "relative",
            "width": "100%",
          }
        }
        tabIndex={0}
        title="Pause Updates"
      >
        <svg
          aria-labelledby="ariaLabelledby"
          className="sc-bwzfXH flKpft"
          version="1.1"
          viewBox="0 0 24 24"
          xmlns="http://www.w3.org/2000/svg"
          xmlnsXlink="http://www.w3.org/1999/xlink"
        >
          <g
            className="sc-bdVaJa iPrjqu"
            fill="currentColor"
            title=""
          >
            <title>
              
            </title>
            <g
              className="glyph"
              fill="rgb(10, 171, 42)"
              transform="translate(-24 -24) scale(3)"
            >
              <path
                d="M6,5 L10.6153846,5 L10.6153846,19 L6,19 L6,5 Z M13.3846154,5 L18,5 L18,19 L13.3846154,19 L13.3846154,5 Z"
                id="Combined-Shape"
              />
            </g>
          </g>
        </svg>
        <span
          className="label"
        >
          
          Pause Updates
          
        </span>
      </button>
    </div>
    <div
      className="sc-htoDjs guaXXs"
      id="ctrl-slider-Instance Polling"
    >
      <div
        aria-disabled={false}
        className="input-range"
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onTouchStart={[Function]}
      >
        <span
          className="input-range__label input-range__label--min"
        >
          <span
            className="input-range__label-container"
          >
            5
          </span>
        </span>
        <div
          className="input-range__track input-range__track--background"
          onMouseDown={[Function]}
          onTouchStart={[Function]}
        >
          <div
            className="input-range__track input-range__track--active"
            style={
              Object {
                "left": "0%",
                "width": "0%",
              }
            }
          />
          <span
            className="input-range__slider-container"
            style={
              Object {
                "left": "0%",
                "position": "absolute",
              }
            }
          >
            <span
              className="input-range__label input-range__label--value"
            >
              <span
                className="input-range__label-container"
              >
                5
              </span>
            </span>
            <div
              aria-controls={undefined}
              aria-labelledby={undefined}
              aria-valuemax={120}
              aria-valuemin={5}
              aria-valuenow={5}
              className="input-range__slider"
              draggable="false"
              onKeyDown={[Function]}
              onMouseDown={[Function]}
              onTouchStart={[Function]}
              role="slider"
              tabIndex="0"
            />
          </span>
        </div>
        <span
          className="input-range__label input-range__label--max"
        >
          <span
            className="input-range__label-container"
          >
            120
          </span>
        </span>
      </div>
      <span
        className="label"
        id="interval-name-Instance Polling"
      >
        Updates Per Minute
      </span>
    </div>
  </div>
</section>
`;
